### PR TITLE
adding a method to retrieve pitch, roll and heading from a quaternion

### PR DESCRIPTION
**Pull request by kaktus40**
Hello,
is it possible to add a little function to retrieve heading, pitch and roll from a quaternion?

### CODE REVIEW

**Comment by pjcozzi** [comment]
Thanks @kaktus40.

File: Source/Core/Quaternion.js [hunk 1]
```diff
@@ -1081,6 +1081,34 @@ define([
...
+     * @returns {Object} an object that contains three attributs (in radians): roll,pitch and heading.
+     */
+    Quaternion.toHeadingPitchRoll = function(quaternion) {
+        //>>includeStart('debug', pragmas.debug);
+        if (!defined(quaternion)) {
+            throw new DeveloperError('quaternion is required');
+        }
+        //>>includeEnd('debug');
+
+        var result = {};
```
Comment (pjcozzi) [inline Source/Core/Quaternion.js]: We generally don't use implicit types like this throughout Cesium.  I think the way we would do this is to create a new type, `HeadingPitchRoll`, that has `heading`, `pitch`, and `roll` properties, and a `fromQuaternion` function.  We may also add this new type to existing Cesium APIs that take heading/pitch/roll, but that doesn't need to be part of this pull request.

**Comment by kaktus40** [comment]
I follow your recommandations

**Comment by pjcozzi** [comment]
Thanks @kaktus40.  I will tweak this a bit and then merge.  We are going to FOSS4G NA next week so it will be later in the month.

**Comment by kaktus40** [comment]
Hello,
I'd like adding a new transform which will be more compliant with the use of aircrafts. Lots of aircraft models follow the described [here](http://en.wikipedia.org/wiki/Aircraft_principal_axes).  actually, a workaround is to use Transforms.headingPitchRollQuaternion with a heading less 90 degrees. With Transforms.headingPitchRollQuaternionAircraft, you don't need the substraction.

**Comment by pjcozzi** [comment]
@kring do you happen to have any spare cycles to review this and perhaps suggest where else we can use `HeadingPitchRoll` in the API?

**Comment by kring** [comment]
@kaktus40 can you update the documentation for the `headingPitchRoll...` functions (both new and old) to indicate the implied axes of the aircraft/camera that is being rotated?  Looking at the code, it looks like `headingPitchRollToFixedFrameAircraft` is measuring heading from West, which seems wrong, but I'm guessing it's actually a difference in which axis is aligned with the nose of the aircraft.  More explicit documentation of the assumptions would make it easier for me (and users) to understand what's going on here.

File: Source/Core/HeadingPitchRoll.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        defaultValue,
+        defined,
+        DeveloperError,
+        freezeObject,
+        CesiumMath,
+        defineProperties) {
+    "use strict";
+
+    /**
+     * heading pitch roll type. Heading is the rotation about the
```
Comment (kring) [inline Source/Core/HeadingPitchRoll.js]: Use complete sentences.  Maybe "A rotation expressed as a heading, pitch, and roll."

File: Source/Core/HeadingPitchRoll.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+     * @param {Number} [heading=0.0] The heading component in radians.
+     * @param {Number} [pitch=0.0] The pitch component in radians.
+     * @param {Number} [roll=0.0] The roll component in radians.
+     */
+    var HeadingPitchRoll = function (heading, pitch, roll) {
+        this._heading = 0.0;
+        this._pitch = 0.0;
+        this._roll = 0.0;
+
+        this.heading = defaultValue(heading, 0.0);
```
Comment (kring) [inline Source/Core/HeadingPitchRoll.js]: These properties will override the ones defined on the prototype below.  Probably not what you want.

But really, the usual approach in Cesium is to allow any value (even invalid ones) to be set for primitive types, and validate the values on use where appropriate.  For example, see the `Spherical` type.  So to keep the same pattern here, the properties defined on the prototype can be removed entirely, along with the `_heading`, etc. properties defined above.

File: Source/Core/HeadingPitchRoll.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    HeadingPitchRoll.fromQuaternion = function (quaternion, result) {
+        //>>includeStart('debug', pragmas.debug);
+        if (!defined(quaternion)) {
+            throw new DeveloperError('quaternion is required');
+        }
+        //>>includeEnd('debug');
+        result = defaultValue(result, new HeadingPitchRoll());
+        var test = 2 * (quaternion.w * quaternion.y - quaternion.z * quaternion.x);
+        var denumRoll = 1 - 2 * (quaternion.x * quaternion.x + quaternion.y * quaternion.y);
+        var numRoll = 2 * (quaternion.w * quaternion.x + quaternion.y * quaternion.z);
```
Comment (kring) [inline Source/Core/HeadingPitchRoll.js]: Avoid abbreviations.  `numeratorRoll` and `denominatorRoll`.

File: Source/Core/HeadingPitchRoll.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                value %= (2 * CesiumMath.PI);
+                this._roll = value;
+            }
+        }
+    });
+
+    /**
+     * Computes the heading, pitch and roll from a quaternion (see http://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles )
+     *
+     * @param {Quaternion} quaternion The quaternion to retrieve heading,pitch and roll in radians.
```
Comment (kring) [inline Source/Core/HeadingPitchRoll.js]: Revise to "The quaternion from which to retrieve heading, pitch, and roll, all expressed in radians."

File: Source/Core/HeadingPitchRoll.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+     * @param {Quaternion} [result] The object onto which to store the result.
+     * @returns {HeadingPitchRoll} The modified result parameter or a new HeadingPitchRoll instance if none was provided retrieving from the quaternion in parameter
```
Comment (kring) [inline Source/Core/HeadingPitchRoll.js]: > The object onto which to store the result.

The object in which to store the result.  If not provided, a new instance is created and returned.

> The modified result parameter or a new HeadingPitchRoll instance if none was provided retrieving from the quaternion in parameter

The modified result parameter or a new HeadingPitchRoll instance if one was not provided.

File: Source/Core/HeadingPitchRoll.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+        result.heading = -Math.atan2(numHeading, denumHeading);
+        result.roll = Math.atan2(numRoll, denumRoll);
+        result.pitch = -Math.asin(test);
+
+        return result;
+    };
+
+    /**
+     * generate a HeadingPitchRoll instance from angles in degrees.
```
Comment (kring) [inline Source/Core/HeadingPitchRoll.js]: Returns a new HeadingPitchRoll instance from angles given in degrees.

File: Source/Core/HeadingPitchRoll.js [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+     *
+     * @param {Number} heading the heading in degrees
+     * @param {Number} pitch the pitch in degrees
+     * @param {Number} roll the heading in degrees
```
Comment (kring) [inline Source/Core/HeadingPitchRoll.js]: Start sentences with capital letters, end them with a period.

File: Source/Core/HeadingPitchRoll.js [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+     * @returns {HeadingPitchRoll} A new HeadingPitchRoll instance
+     */
+    HeadingPitchRoll.fromDegrees = function (heading, pitch, roll) {
```
Comment (kring) [inline Source/Core/HeadingPitchRoll.js]: Allow the user to specify an optional result parameter.

File: Source/Core/Quaternion.js [hunk 10]
```diff
@@ -1081,7 +1081,7 @@ define([
                 (Math.abs(left.z - right.z) <= epsilon) &&
                 (Math.abs(left.w - right.w) <= epsilon));
     };
-
+
```
Comment (kring) [inline Source/Core/Quaternion.js]: Please revert this unnecessary whitespace change.

File: Source/Core/Transforms.js [hunk 11]
```diff
@@ -380,6 +476,36 @@ define([
...
+     *
+     * @example
+     * // Get the transform from local heading-pitch-roll at cartographic (0.0, 0.0) to Earth's fixed frame.
+     * var center = Cesium.Cartesian3.fromDegrees(0.0, 0.0);
+     * var heading = -Cesium.Math.PI_OVER_TWO;
+     * var pitch = Cesium.Math.PI_OVER_FOUR;
+     * var roll = 0.0;
+     * var transform = Cesium.Transforms.headingPitchRollToFixedFrameAircraft(center, heading, pitch, roll);
+     */
+    Transforms.headingPitchRollToFixedFrameAircraft = function(origin, heading, pitch, roll, ellipsoid, result) {
```
Comment (kring) [inline Source/Core/Transforms.js]: Maybe name this `aircraftHeadingPitchRollToFixedFrame`?

File: Specs/Core/HeadingPitchRollSpec.js [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+], function (
+        HeadingPitchRoll,
+        Quaternion,
+        CesiumMath
+        ) {
+    "use strict";
+    /*global it,expect*/
+
+    it('conversion from quaternion', function () {
+        var deg2rad = Math.PI / 180;
```
Comment (kring) [inline Specs/Core/HeadingPitchRollSpec.js]: Use the `CesiumMath` value.

File: Specs/Core/QuaternionSpec.js [hunk 13]
```diff
@@ -1024,6 +1024,6 @@ defineSuite([
             Quaternion.fastSquad(new Quaternion(), new Quaternion(), new Quaternion(), new Quaternion(), 3);
         }).toThrowDeveloperError();
     });
-
+
```
Comment (kring) [inline Specs/Core/QuaternionSpec.js]: Please revert this unnecessary whitespace change, too.

**Comment by kaktus40** [comment]
@kring  : Hello I follow your recommendations and correct the codes. Sorry for my bad English.  About your first comment, I think Transforms.northWestUpToFixedFrame (that I create) should be a Transforms.northEastDownToFixedFrame to be compliant with aircraft referential (abscissas axis is the line from queue to nose, ordered axis is the line from left to right  when we see an aircraft from above and z axis is from above to below the aircraft).
What do you think?

**Comment by kring** [comment]
Thanks @kaktus40.  I haven't looked in detail at your new changes yet, but I wanted to respond to your comment since I've already let 4 days get behind me...

>  I think Transforms.northWestUpToFixedFrame (that I create) should be a Transforms.northEastDownToFixedFrame to be compliant with aircraft referential

That sounds confusing to me, assuming I understand what you mean, because the concept of a "north-east-down" reference frame doesn't have any connection to aircraft at all.  It's simply: X points North, Y points East, Z points down.

Something like `Transforms.aircraftPrincipalFrameToFixedFrame` could be ok.  Then the documentation defines "aircraft principal frame" as a reference frame in which the X axis, which is directed from the aircraft tail to its nose, points North, the Y axis, which is directed directed from the left wing to the right wing, points East, and the Z axis points down.  But what I've just described is identical to the actual North-East-Down frame, so presumably I have it wrong.

**Comment by kaktus40** [comment]
Hello, it's my mistake, I didn't see there was a northEastDown conversion. So I don't need Transforms.northWestUpToFixedFrame. The aircraft quaternion should work with northEastDown transform. I'll make the correction. Thanks for the review

**Comment by kaktus40** [comment]
Hello, I remove Transform.northWestUpToFixedFrame and use Transforms.northEastDownToFixedFrame in order to make Transforms.aircraftHeadingPitchRollToFixedFrame

**Comment by pjcozzi** [comment]
@kring @kaktus40 what is the plan with this pull request?

**Comment by hpinkos** [comment]
@kring What's the story here?  People on the forum ask for a function to do this every now and again.

**Comment by pjcozzi** [comment]
@kaktus40 @kring  I think this is a valuable contribution for Cesium.  Is it possible for us to converge on an approach and merge this?

**Comment by kaktus40** [comment]
Hello,
maybe I should dissociate this PR in two:
 - one PR to retrieve pitch, roll and heading from a quaternion ?
 - one PR to propose aircraftHeadingPitchRollToFixedFrame ?

**Comment by pjcozzi** [comment]
Sounds like a good plan, thanks @kaktus40!
